File: src/main/java/org/entities/passive/Isolsi.java
package org.entities.passive;

import org.engine.Main;
import org.entities.Autonomous;
import org.graphics.Graphics;
import org.graphics.Timer;
import org.input.Keyboard;
import org.loader.ImageResource;
import org.loader.ResourceHandler;
import org.world.World;

import static com.jogamp.newt.event.KeyEvent.VK_SPACE;

@SuppressWarnings("StatementWithEmptyBody")
public class Isolsi extends Autonomous {
    private static final int LOOK_DOWN=3, GESTURE=4, SLIGHT_ANGER=5, CLENCH_FIST=6, POINTING=7, ANGRY=8, QUESTION=9;
    private int convoState=0;
    private boolean crossfadeActive, inPostLaranoScene;
    private Timer crossfade=new Timer(0,1,1,0.01f,60);
    private ImageResource[] isolsi= ResourceHandler.getGolemLoader().getIsolsi();
    private String[] postLaranoConvo ={"That orange golem certainly caused you a bit of trouble didn't he.",
            "Does he not know the importance of your mission?!",
            "The nerve of those lesser golems!",
            "If anyone should try to stand in our way like that again...",
            "don't hold back."};
    private String[] postBoulderConvo={"What are you doing?",
            "You could have died doing that!",
            "And for what?",
            "The precious model town is safe...",
            "It's not real!",
            "It's not worth our time!",
            "Now stop fooling around...",
            "...and TAKE THINGS SERIOUSLY!"};
    
    public Isolsi(boolean inPostLaranoScene) {
        super(3, 64, 60);
        if(!inPostLaranoScene){
            y=7;
            subLevel=5;
            state=2;
        }
        this.inPostLaranoScene =inPostLaranoScene;
        reset();
    }

    @Override
    public void update() {
        if(inPostLaranoScene)larano();
        else boulder();
        y+=vY;
        vY-=vY-World.getGravity()>=0?World.getGravity():vY;
        crossfade.setActive(crossfadeActive);
        crossfade.update();
    }

    private void larano(){
        switch(state){
            case -1:
                if(Main.getHarold().getX()>15){
                    Main.getHarold().setLockControls(true);
                    state++;
                }
                break;
            case 0:
                if(y>7)vY=-0.15f;
                else{
                    crossfadeActive=true;
                }
                if(crossfade.isActive()&&crossfade.getCurrent()==0){
                    crossfade.setCurrent(1);
                    state++;
                }
                break;
            case 1:
                if(crossfade.isActive()&&crossfade.getCurrent()==0){
                    crossfadeActive=false;
                    crossfade.setCurrent(1);
                    state++;
                }
                break;
            default:
                if(Keyboard.keys.contains(VK_SPACE)&&convoState+1< postLaranoConvo.length){
                    convoState++;
                }else if(Keyboard.keys.contains(VK_SPACE)){
                    World.setLevelTransition(true);
                }
                while(Keyboard.keys.contains(VK_SPACE)){}
        }
    }

    private void boulder(){
        Main.getHarold().setX(10);
        ResourceHandler.getHaroldLoader().setDirection(true);
        if(state==2||convoState==0) {
            Main.getHarold().setLockControls(true);
            if (Keyboard.keys.contains(VK_SPACE) && convoState + 1 < postBoulderConvo.length) {
                convoState++;
            } else if (Keyboard.keys.contains(VK_SPACE)) {
                state++;
            }
            while (Keyboard.keys.contains(VK_SPACE)) {
            }
        }else if(state==3){
            World.setMasterColor(1,1,1);
            World.getMaster().setDirection(false);
            if(World.getMaster().getCurrent()==0){
                World.incrementSubLevel();
                World.getMaster().setDirection(true);
                Main.getHarold().setLockControls(false);
            }
        }
    }

    @Override
    public void render() {
        if (state == -1) return;
        if (state < 2){
            Graphics.setDrawColor(1, 1, 1, crossfade.getCurrent());
            Graphics.drawImage(isolsi[state], x - Graphics.toWorldWidth(calcXOffset(state)),
                    y + Graphics.convertToWorldHeight(calcYOffset(state)));
            if (crossfadeActive && state + 1 < isolsi.length) {
                Graphics.setDrawColor(1, 1, 1, 1 - crossfade.getCurrent());
                Graphics.drawImage(isolsi[state + 1],
                        x - Graphics.toWorldWidth(calcXOffset(state + 1)),
                        y + Graphics.convertToWorldHeight(calcYOffset(state + 1)));
            }
        }else{
            Graphics.drawImage(isolsi[calcConvoSprite()],
                    x - Graphics.toWorldWidth(calcXOffset(calcConvoSprite())),
                    y + Graphics.convertToWorldHeight(calcYOffset(calcConvoSprite())));
        }
        Graphics.setDrawColor(1,1,1,1);
        Graphics.setFont(Graphics.SMALL);
        if(convoState>-1){
            if(inPostLaranoScene)Graphics.drawText(postLaranoConvo[convoState],45,45,20,true);
            else Graphics.drawText(postBoulderConvo[convoState],45,45,20,true);
        }
    }

    private float calcXOffset(int frame){
        switch(frame){
            case 0: return -26;
            case 1: return 91-37;
            case 2: //Collapse
            case LOOK_DOWN: return 54-37;
            case GESTURE: return 106-37;
            case SLIGHT_ANGER: return 139-37;
            case CLENCH_FIST: return 92-37;
            case POINTING: return 129-37;
            case ANGRY: return 113-37;
            case QUESTION: return 106-37;
            default: return 0;
        }
    }

    private float calcYOffset(int frame){
        switch(frame){
            case 0: return 387;
            case 1: return -35.8f;
            default: return 0;
        }
    }

    private int calcConvoSprite(){
        if(inPostLaranoScene)
            switch(convoState){
                case 4: return CLENCH_FIST;
                case 1: return GESTURE;
                case 2: return SLIGHT_ANGER;
                case 3:
                case 0:
                default:return LOOK_DOWN;
            }
        else
            switch(convoState){
                case 0: return QUESTION;
                case 1: return SLIGHT_ANGER;
                case 3: return GESTURE;
                case 4:
                case 5:
                case 2: return ANGRY;
                case 6:
                case 7: return POINTING;
                default: return LOOK_DOWN;
            }
    }

    @Override
    public void reset() {
        if(inPostLaranoScene)state=-1;
        else state=2;
        convoState=0;
        crossfadeActive=false;
        crossfade.setActive(false);
        crossfade.setDirection(false);
        crossfade.setCurrent(1);
    }
}

File: src/main/java/org/entities/passive/Isolsi.kt
package org.entities.passive

import com.jogamp.newt.event.KeyEvent
import org.engine.Main
import org.entities.Autonomous
import org.graphics.Graphics
import org.graphics.TimedGradient
import org.input.Keyboard
import org.loader.ResourceHandler
import org.world.World.gravity
import org.world.World.incrementSubLevel
import org.world.World.master
import org.world.World.setLevelTransition
import org.world.World.setMasterColor

class Isolsi(inPostLaranoScene: Boolean) : Autonomous(3, 64f, 60f) {
    private var convoState = 0
    private var crossfadeActive = false
    private val inPostLaranoScene: Boolean
    private val crossfade = TimedGradient(0f, 1f, 1f, 0.01f, 60)
    private val isolsi = ResourceHandler.getGolemLoader().isolsi
    private val postLaranoConvo = arrayOf(
        "That orange golem certainly caused you a bit of trouble didn't he.",
        "Does he not know the importance of your mission?!",
        "The nerve of those lesser golems!",
        "If anyone should try to stand in our way like that again...",
        "don't hold back."
    )
    private val postBoulderConvo = arrayOf(
        "What are you doing?",
        "You could have died doing that!",
        "And for what?",
        "The precious model town is safe...",
        "It's not real!",
        "It's not worth our time!",
        "Now stop fooling around...",
        "...and TAKE THINGS SERIOUSLY!"
    )

    override fun update(deltaTime: Float) {
        if (inPostLaranoScene)
            larano()
        else
            boulder()
        y += vY
        vY -= if (vY - gravity >= 0) gravity else vY
        crossfade.isActive = crossfadeActive
        crossfade.update()
    }

    private fun larano() {
        when (state) {
            -1 -> if (Main.getHarold().x > 15) {
                Main.getHarold().setLockControls(true)
                state++
            }
            0 -> {
                if (y > 7) vY = -0.15f else {
                    crossfadeActive = true
                }
                if (crossfade.isActive && crossfade.current == 0f) {
                    crossfade.current = 1f
                    state++
                }
            }
            1 -> if (crossfade.isActive && crossfade.current == 0f) {
                crossfadeActive = false
                crossfade.current = 1f
                state++
            }
            else -> {
                if (Keyboard.keys.contains(KeyEvent.VK_SPACE) && convoState + 1 < postLaranoConvo.size) {
                    convoState++
                } else if (Keyboard.keys.contains(KeyEvent.VK_SPACE)) {
                    setLevelTransition(true)
                }
                while (Keyboard.keys.contains(KeyEvent.VK_SPACE)) {
                }
            }
        }
    }

    private fun boulder() {
        Main.getHarold().x = 10f
        ResourceHandler.getHaroldLoader().setDirection(true)
        if (state == 2 || convoState == 0) {
            Main.getHarold().setLockControls(true)
            if (Keyboard.keys.contains(KeyEvent.VK_SPACE) && convoState + 1 < postBoulderConvo.size) {
                convoState++
            } else if (Keyboard.keys.contains(KeyEvent.VK_SPACE)) {
                state++
            }
            while (Keyboard.keys.contains(KeyEvent.VK_SPACE)) {
            }
        } else if (state == 3) {
            setMasterColor(1f, 1f, 1f)
            master.direction = false
            if (master.current == 0f) {
                incrementSubLevel()
                master.direction = true
                Main.getHarold().setLockControls(false)
            }
        }
    }

    override fun render() {
        if (state == -1) return
        if (state < 2) {
            Graphics.setDrawColor(1f, 1f, 1f, crossfade.current)
            Graphics.drawImage(
                isolsi[state], x - Graphics.toWorldWidth(calcXOffset(state)),
                y + Graphics.toWorldHeight(calcYOffset(state))
            )
            if (crossfadeActive && state + 1 < isolsi.size) {
                Graphics.setDrawColor(1f, 1f, 1f, 1 - crossfade.current)
                Graphics.drawImage(
                    isolsi[state + 1],
                    x - Graphics.toWorldWidth(calcXOffset(state + 1)),
                    y + Graphics.toWorldHeight(calcYOffset(state + 1))
                )
            }
        } else {
            Graphics.drawImage(
                isolsi[calcConvoSprite()],
                x - Graphics.toWorldWidth(calcXOffset(calcConvoSprite())),
                y + Graphics.toWorldHeight(calcYOffset(calcConvoSprite()))
            )
        }
        Graphics.setDrawColor(1f, 1f, 1f, 1f)
        Graphics.setFont(Graphics.FontType.SMALL)
        if (convoState > -1) {
            if (inPostLaranoScene) {
                Graphics.drawText(postLaranoConvo[convoState], 45f, 45f, 20f, true)
            } else {
                Graphics.drawText(postBoulderConvo[convoState], 45f, 45f, 20f, true)
            }
        }
    }

    private fun calcXOffset(frame: Int): Float {
        return when (frame) {
            0 -> (-26).toFloat()
            1 -> (91 - 37).toFloat()
            2, LOOK_DOWN -> (54 - 37).toFloat()
            GESTURE -> (106 - 37).toFloat()
            SLIGHT_ANGER -> (139 - 37).toFloat()
            CLENCH_FIST -> (92 - 37).toFloat()
            POINTING -> (129 - 37).toFloat()
            ANGRY -> (113 - 37).toFloat()
            QUESTION -> (106 - 37).toFloat()
            else -> 0f
        }
    }

    private fun calcYOffset(frame: Int): Float {
        return when (frame) {
            0 -> 387f
            1 -> -35.8f
            else -> 0f
        }
    }

    private fun calcConvoSprite(): Int {
        return if (inPostLaranoScene) when (convoState) {
            4 -> CLENCH_FIST
            1 -> GESTURE
            2 -> SLIGHT_ANGER
            3, 0 -> LOOK_DOWN
            else -> LOOK_DOWN
        } else when (convoState) {
            0 -> QUESTION
            1 -> SLIGHT_ANGER
            3 -> GESTURE
            4, 5, 2 -> ANGRY
            6, 7 -> POINTING
            else -> LOOK_DOWN
        }
    }

    override fun reset() {
        state = if (inPostLaranoScene) -1 else 2
        convoState = 0
        crossfadeActive = false
        crossfade.isActive = false
        crossfade.direction = false
        crossfade.current = 1f
    }

    companion object {
        private const val LOOK_DOWN = 3
        private const val GESTURE = 4
        private const val SLIGHT_ANGER = 5
        private const val CLENCH_FIST = 6
        private const val POINTING = 7
        private const val ANGRY = 8
        private const val QUESTION = 9
    }

    init {
        if (!inPostLaranoScene) {
            y = 7f
            subLevel = 5
            state = 2
        }
        this.inPostLaranoScene = inPostLaranoScene
        reset()
    }
}